### PR TITLE
New export API with dynamic shape specifications instead of constraints

### PR DESCRIPTION
**Pull request by avikchaudhuri**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108448

Our experience using `constraints` / `dynamic_dim` with the existing export API has found it to be (subjectively) clunky and (objectively) verbose in common cases.

This PR implements a new design for the export API that replaces the use of `constraints` / `dynamic_dim` with a new way of specifying dynamic shapes, involving the following concepts: 
* a constructor `Dim` for first-class named dynamic dimensions with ranges (similar to `functorch.dim`, and analogous to internal symbolic sizes)
* a mechanism that uses the above in `export` calls to associate inputs to their dynamic shape specifications (`dynamic_shapes`)

Design doc: https://docs.google.com/presentation/d/168U7XK72C_WSsZpGESP6Cho9udh193fi0gfjxCNcJ4E/edit#slide=id.p (Meta-only). Note that we only implement Option 1 in that doc. An older version of this PR also implemented Option 3, which is an alternative way of specifying dynamic shapes using tensor type annotations on the exported callable; but we have moved that to future work for now.

See docs for these new features in `torch.export`. The existing `torch.export.export` is modified to use the new API, `torch._export.export__RC__`, whenever `constraints=None`. We have not deprecated the existing API yet, but will do in a follow-up.

Constraint violation errors arising through use of the new API will now contain suggested fixes using the new API. No longer do we need to report all specializations for static dimensions and suggest all constraints over dynamic dimensions to fix such errors. Instead, due to the redesign, the suggested fixes are much more concise, only involving modifying the definitions of relevant `Dim`s.

Differential Revision: [D48919204](https://our.internmc.facebook.com/intern/diff/D48919204/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_export/__init__.py [hunk 1]
```diff
@@ -62,7 +65,157 @@
...
+
+def Dim(name, *, min=None, max=None):
+    """
+    Constructs a type analogous to a named symbolic integer with a range.
+    It can be used to describe multiple possible values of a tensor dimension.
+    Note that different dimensions of the same tensor, or of different tensors,
+    can be described by the same type.
+    """
+    _min = 2 if min is None else min
+    _max = sys.maxsize if max is None else max
```
Comment (gmagogsfm) [inline torch/_export/__init__.py:83]:
```suggestion
    _min = 2 if min is None else builtins.max(min, 2)
    _max = sys.maxsize if max is None else builtins.max(max, 2)
    assert _max > _min
```

File: torch/_export/__init__.py [hunk 2]
```diff
@@ -62,7 +65,157 @@
...
+    return _Dim(name, (int,), {"min": _min, "max": _max})
+
+
+def dims(*names: str):
```
Comment (gmagogsfm) [inline torch/_export/__init__.py:87]: Just making sure I understand this correctly: users are not supposed to modify min/max of created Dim objects, and dims doesn't take max/min arguments, so `dims` can only be used to attach names to dims?

File: torch/export/__init__.py [hunk 3]
```diff
@@ -693,12 +705,18 @@ def __eq__(self, other):
             vr=self.constraint_range.vr & other.constraint_range.vr,
             warn_only=False,
         )
+        if self.debug_name is None:
+            debug_name = other.debug_name
+        else:
+            assert other.debug_name is None or self.debug_name == other.debug_name
```
[the remark this answers is not in the document]
Comment (avikchaudhuri) [reply torch/export/__init__.py:711]: This only guards against inconsistent debug names. With the old API debug names will always be `None`.

File: torch/_export/__init__.py [hunk 4]
```diff
@@ -62,7 +65,157 @@
...
+        elif isinstance(shape, dict):
+            for i, dim in shape.items():
+                if isinstance(dim, _Dim):
+                    symbols[dim.__name__].append(dynamic_dim(tensor, i, debug_name=dim.__name__))
+
+    import inspect
+    signature = inspect.signature(f.forward) if isinstance(f, torch.nn.Module) else inspect.signature(f)
+    combined_args = signature.bind(*args, **kwargs).arguments
+
+    if dynamic_shapes is None:
```
[the remark this answers is not in the document]
Comment (avikchaudhuri) [reply torch/_export/__init__.py:198]: Yes.

File: torch/_export/__init__.py (hunk 2 above)
Comment (avikchaudhuri) [reply torch/_export/__init__.py:87]: Sorry, this is an oversight. I should add min/max args here.

File: torch/export/__init__.py [hunk 5]
```diff
@@ -1020,9 +1042,12 @@ def specify_constraints(x):
...
     """
 
-    from torch._export import export
+    from torch._export import export, export__RC__
 
-    return export(f, args, kwargs, constraints)
+    if constraints is not None:
+        return export(f, args, kwargs, constraints)
+    else:
+        return export__RC__(f, args, kwargs, dynamic_shapes=dynamic_shapes)
```
[the remark this answers is not in the document]
Comment (avikchaudhuri) [reply torch/export/__init__.py:1050]: I'll punt on it. Otherwise I'll have to check annotations right here. Given we will deprecate or even remove `constraints` soon, I doubt this will come up in practice.

File: torch/fx/experimental/symbolic_shapes.py [hunk 6]
```diff
@@ -1960,7 +1966,103 @@ def remove_redundant_dynamic_results(self):
...
+
+            results = defaultdict(dict)
+            for s in self._static_results.union(self._dynamic_results):
+                t = transform(s)
+                if t == s:
+                    continue
+                left, op, right = t.split(" ")
+                if op == "==" and left == right:
+                    continue
+                if left.isdigit():
```
[the remark this answers is not in the document]
Comment (avikchaudhuri) [reply torch/fx/experimental/symbolic_shapes.py:1989]: `op` needs to be flipped too

[the remark this answers is not in the document]
Comment (gmagogsfm) [reply torch/fx/experimental/symbolic_shapes.py:1989]: Yeah, replacing all "<" with ">" and vice versa should be enough.

**Review by suo — CHANGES_REQUESTED** [review]

File: torch/_export/__init__.py [hunk 7]
```diff
@@ -62,7 +66,204 @@
...
+    Here, `dynamic_shapes` is expected to be a (possibly partial) dict from
+    argument names of `f` to dynamic shape specifications, as follows:
+    - The dynamic shape of a tensor argument can be specified as a dict from
+      dynamic dimension indices to Dim types. It is not required to include
+      static dimension indices in this dict, but when they are, they should
+      be mapped to _.
+    - Arguments that are dicts or tuples of tensors are recursively specified
+      by using mappings or sequences of contained specifications.
+
+    Mechanism #2: types
```
[the remark this answers is not in the document]
Comment (avikchaudhuri) [reply torch/_export/__init__.py:180]: > First, there should ideally be one way to do things. The type annotation mechanism seems strictly less powerful (e.g. it seems impossible to express something like a dict where specific keys have different constraints). I think if we want two mechanisms, there needs to be a strong reason.

I agree that one way is better than two, but I think the call-site and the function-site use cases cater to different sets of users.

Actually there is enough power in the type annotation mechanism right now to match whatever you can pass to `dynamic_shapes`. In particular I've included a `dynamic_shapes_type` function to map one to the other. For the case you mention, you'd use a `TypedDict`.

That said I'm somewhat frustrated that a lot of good things in `typing` happened only in 3.9+ and so I've to resort to less-than-ideal implementations to remain compatible with 3.8.

> Second, static type annotations come with the user expectation that the annotation is universally true, e.g. if I type a function def foo(i: int), I come in with the expectation that i will always be an int, regardless of how foo is called. This is not true of these type annotations—they are only respected if the function happens to be the top-level exported function. I can see this being potentially quite confusing.

I agree. Coupled with the fact that I'm not quite happy with the way the type annotations are encoded right now, this feels like a good reason to take it out.

I do absolutely want to bring it back at some point. So I might have to leave in some tests to ensure that in principle the two mechanisms are equally expressive.

File: torch/_export/__init__.py [hunk 8]
```diff
@@ -62,7 +66,204 @@
...
+      and static, must be included.
+    - Arguments that are dicts or tuples of tensors are recursively typed
+      by using Dict[str, ...] or List[...] of contained specifications.
+    """
+    kwargs = kwargs if kwargs is not None else {}
+
+    from collections.abc import Mapping, Sequence
+    from typing import get_origin, get_args, _TypedDictMeta  # type: ignore[attr-defined]
+
+    def assoc_zip(combined_args, dynamic_shapes):
```
[the remark this answers is not in the document]
Comment (avikchaudhuri) [reply torch/_export/__init__.py:130]: Handling dataclasses in the most recent version. 

I would have used `pytree` if there was something like `tree_zip` (something that descends into the same structure of multiple things at once). I think having this custom descent is fine for now.

Key paths: yeah, maybe good for partial specs later. For now it is convenient enough to `pytree.tree_map` over inputs to derive specs.

File: torch/_export/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gmagogsfm) [reply torch/_export/__init__.py:164]: Should this part of logic be removed now that we don't have TensorType?

**Review by gmagogsfm — APPROVED** [review]
Looks good, just some minor nits

File: torch/_export/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (avikchaudhuri) [reply torch/_export/__init__.py:164]: Oh yes.

File: torch/_export/__init__.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (suo) [reply torch/_export/__init__.py:133]: yeah agree, it's just an implementation detail. We can improve in the future

**Review by suo — APPROVED** [review]
lgtm!
